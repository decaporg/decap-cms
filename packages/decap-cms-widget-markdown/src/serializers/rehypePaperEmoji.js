/**
 * Dropbox Paper outputs emoji characters as images, and stores the actual
 * emoji character in a `data-emoji-ch` attribute on the image. This plugin
 * replaces the images with the emoji characters.
 */
export default function rehypePaperEmoji() {
  function transform(node) {
    if (node.tagName === 'img' && node.properties.dataEmojiCh) {
      return { type: 'text', value: node.properties.dataEmojiCh };
    }
    node.children = node.children ? node.children.map(transform) : node.children;
    return node;
  }

  return transform;
}
