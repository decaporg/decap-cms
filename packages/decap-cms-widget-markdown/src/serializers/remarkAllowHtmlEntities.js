export default function remarkAllowHtmlEntities() {
  this.Parser.prototype.inlineTokenizers.text = text;

  /**
   * This is a port of the `remark-parse` text tokenizer, adapted to exclude
   * HTML entity decoding.
   */
  function text(eat, value, silent) {
    var self = this;
    var methods;
    var tokenizers;
    var index;
    var length;
    var subvalue;
    var position;
    var tokenizer;
    var name;
    var min;

    /* istanbul ignore if - never used (yet) */
    if (silent) {
      return true;
    }

    methods = self.inlineMethods;
    length = methods.length;
    tokenizers = self.inlineTokenizers;
    index = -1;
    min = value.length;

    while (++index < length) {
      name = methods[index];

      if (name === 'text' || !tokenizers[name]) {
        continue;
      }

      tokenizer = tokenizers[name].locator;

      if (!tokenizer) {
        eat.file.fail('Missing locator: `' + name + '`');
      }

      position = tokenizer.call(self, value, 1);

      if (position !== -1 && position < min) {
        min = position;
      }
    }

    subvalue = value.slice(0, min);

    eat(subvalue)({
      type: 'text',
      value: subvalue,
    });
  }
}
