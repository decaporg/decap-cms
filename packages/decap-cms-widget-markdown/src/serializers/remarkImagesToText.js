/**
 * Images must be parsed as shortcodes for asset proxying. This plugin converts
 * MDAST image nodes back to text to allow shortcode pattern matching. Note that
 * this transformation only occurs for images that are the sole child of a top
 * level paragraph - any other image is left alone and treated as an inline
 * image.
 */
export default function remarkImagesToText() {
  return transform;

  function transform(node) {
    const children = node.children.map(child => {
      if (
        child.type === 'paragraph' &&
        child.children.length === 1 &&
        child.children[0].type === 'image'
      ) {
        const { alt, url, title } = child.children[0];
        const value = `![${alt || ''}](${url || ''}${title ? ` "${title}"` : ''})`;
        child.children = [{ type: 'text', value }];
      }
      return child;
    });
    return { ...node, children };
  }
}
