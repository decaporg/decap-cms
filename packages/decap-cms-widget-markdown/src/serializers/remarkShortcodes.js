export function remarkParseShortcodes({ plugins }) {
  const Parser = this.Parser;
  const tokenizers = Parser.prototype.blockTokenizers;
  const methods = Parser.prototype.blockMethods;

  tokenizers.shortcode = createShortcodeTokenizer({ plugins });

  methods.unshift('shortcode');
}

export function getLinesWithOffsets(value) {
  const SEPARATOR = '\n\n';
  const splitted = value.split(SEPARATOR);
  const trimmedLines = splitted
    .reduce(
      (acc, line) => {
        const { start: previousLineStart, originalLength: previousLineOriginalLength } =
          acc[acc.length - 1];

        return [
          ...acc,
          {
            line: line.trimEnd(),
            start: previousLineStart + previousLineOriginalLength + SEPARATOR.length,
            originalLength: line.length,
          },
        ];
      },
      [{ start: -SEPARATOR.length, originalLength: 0 }],
    )
    .slice(1)
    .map(({ line, start }) => ({ line, start }));
  return trimmedLines;
}

function matchFromLines({ trimmedLines, plugin }) {
  for (const { line, start } of trimmedLines) {
    const match = line.match(plugin.pattern);
    if (match) {
      match.index += start;
      return match;
    }
  }
}

function createShortcodeTokenizer({ plugins }) {
  return function tokenizeShortcode(eat, value, silent) {
    // Plugin patterns may rely on `^` and `$` tokens, even if they don't
    // use the multiline flag. To support this, we fall back to searching
    // through each line individually, trimming trailing whitespace and
    // newlines, if we don't initially match on a pattern. We keep track of
    // the starting position of each line so that we can sort correctly
    // across the full multiline matches.
    const trimmedLines = getLinesWithOffsets(value);

    // Attempt to find a regex match for each plugin's pattern, and then
    // select the first by its occurrence in `value`. This ensures we won't
    // skip a plugin that occurs later in the plugin registry, but earlier
    // in the `value`.
    const [{ plugin, match } = {}] = plugins
      .toArray()
      .map(plugin => ({
        match: value.match(plugin.pattern) || matchFromLines({ trimmedLines, plugin }),
        plugin,
      }))
      .filter(({ match }) => !!match)
      .sort((a, b) => a.match.index - b.match.index);

    if (match) {
      if (silent) {
        return true;
      }

      const shortcodeData = plugin.fromBlock(match);

      try {
        return eat(match[0])({
          type: 'shortcode',
          data: { shortcode: plugin.id, shortcodeData },
        });
      } catch (e) {
        console.warn(
          `Sent invalid data to remark. Plugin: ${plugin.id}. Value: ${
            match[0]
          }. Data: ${JSON.stringify(shortcodeData)}`,
        );
        return false;
      }
    }
  };
}

export function createRemarkShortcodeStringifier({ plugins }) {
  return function remarkStringifyShortcodes() {
    const Compiler = this.Compiler;
    const visitors = Compiler.prototype.visitors;

    visitors.shortcode = shortcode;

    function shortcode(node) {
      const { data } = node;
      const plugin = plugins.find(plugin => data.shortcode === plugin.id);
      return plugin.toBlock(data.shortcodeData);
    }
  };
}
