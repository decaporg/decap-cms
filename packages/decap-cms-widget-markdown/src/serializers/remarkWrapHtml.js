import u from 'unist-builder';

/**
 * Ensure that top level 'html' type nodes are wrapped in paragraphs. Html nodes
 * are used for text nodes that we don't want Remark or Rehype to parse.
 */
export default function remarkWrapHtml() {
  function transform(tree) {
    tree.children = tree.children.map(node => {
      if (node.type === 'html') {
        return u('paragraph', [node]);
      }
      return node;
    });

    return tree;
  }

  return transform;
}
