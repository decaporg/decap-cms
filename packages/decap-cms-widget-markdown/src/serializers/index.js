import trimEnd from 'lodash/trimEnd';
import unified from 'unified';
import u from 'unist-builder';
import markdownToRemarkPlugin from 'remark-parse';
import remarkToMarkdownPlugin from 'remark-stringify';
import remarkToRehype from 'remark-rehype';
import rehypeToHtml from 'rehype-stringify';
import htmlToRehype from 'rehype-parse';
import rehypeToRemark from 'rehype-remark';

import remarkToRehypeShortcodes from './remarkRehypeShortcodes';
import rehypePaperEmoji from './rehypePaperEmoji';
import remarkAssertParents from './remarkAssertParents';
import remarkPaddedLinks from './remarkPaddedLinks';
import remarkWrapHtml from './remarkWrapHtml';
import remarkToSlate from './remarkSlate';
import remarkSquashReferences from './remarkSquashReferences';
import { remarkParseShortcodes, createRemarkShortcodeStringifier } from './remarkShortcodes';
import remarkEscapeMarkdownEntities from './remarkEscapeMarkdownEntities';
import remarkStripTrailingBreaks from './remarkStripTrailingBreaks';
import remarkAllowHtmlEntities from './remarkAllowHtmlEntities';
import slateToRemark from './slateRemark';
import { getEditorComponents } from '../MarkdownControl';

/**
 * This module contains all serializers for the Markdown widget.
 *
 * The value of a Markdown widget is transformed to various formats during
 * editing, and these formats are referenced throughout serializer source
 * documentation. Below is brief glossary of the formats used.
 *
 * - Markdown {string}
 *   The stringified Markdown value. The value of the field is persisted
 *   (stored) in this format, and the stringified value is also used when the
 *   editor is in "raw" Markdown mode.
 *
 * - MDAST {object}
 *   Also loosely referred to as "Remark". MDAST stands for MarkDown AST
 *   (Abstract Syntax Tree), and is an object representation of a Markdown
 *   document. Underneath, it's a Unist tree with a Markdown-specific schema.
 *   MDAST syntax is a part of the Unified ecosystem, and powers the Remark
 *   processor, so Remark plugins may be used.
 *
 * - HAST {object}
 *   Also loosely referred to as "Rehype". HAST, similar to MDAST, is an object
 *   representation of an HTML document.  The field value takes this format
 *   temporarily before the document is stringified to HTML.
 *
 * - HTML {string}
 *   The field value is stringified to HTML for preview purposes - the HTML value
 *   is never parsed, it is output only.
 *
 * - Slate Raw AST {object}
 *   Slate's Raw AST is a very simple and unopinionated object representation of
 *   a document in a Slate editor. We define our own Markdown-specific schema
 *   for serialization to/from Slate's Raw AST and MDAST.
 */

/**
 * Deserialize a Markdown string to an MDAST.
 */
export function markdownToRemark(markdown, remarkPlugins) {
  const processor = unified()
    .use(markdownToRemarkPlugin, { fences: true, commonmark: true })
    .use(markdownToRemarkRemoveTokenizers, { inlineTokenizers: ['url'] })
    .use(remarkParseShortcodes, { plugins: getEditorComponents() })
    .use(remarkAllowHtmlEntities)
    .use(remarkSquashReferences)
    .use(remarkPlugins);

  /**
   * Parse the Markdown string input to an MDAST.
   */
  const parsed = processor.parse(markdown);

  /**
   * Further transform the MDAST with plugins.
   */
  const result = processor.runSync(parsed);

  return result;
}

/**
 * Remove named tokenizers from the parser, effectively deactivating them.
 */
function markdownToRemarkRemoveTokenizers({ inlineTokenizers }) {
  inlineTokenizers &&
    inlineTokenizers.forEach(tokenizer => {
      delete this.Parser.prototype.inlineTokenizers[tokenizer];
    });
}

/**
 * Serialize an MDAST to a Markdown string.
 */
export function remarkToMarkdown(obj, remarkPlugins) {
  /**
   * Rewrite the remark-stringify text visitor to simply return the text value,
   * without encoding or escaping any characters. This means we're completely
   * trusting the markdown that we receive.
   */
  function remarkAllowAllText() {
    const Compiler = this.Compiler;
    const visitors = Compiler.prototype.visitors;
    visitors.text = node => node.value;
  }

  /**
   * Provide an empty MDAST if no value is provided.
   */
  const mdast = obj || u('root', [u('paragraph', [u('text', '')])]);

  const remarkToMarkdownPluginOpts = {
    commonmark: true,
    fences: true,
    listItemIndent: '1',

    /**
     * Use asterisk for everything, it's the most versatile. Eventually using
     * other characters should be an option.
     */
    bullet: '*',
    emphasis: '*',
    strong: '*',
    rule: '-',
  };

  const processor = unified()
    .use({ settings: remarkToMarkdownPluginOpts })
    .use(remarkEscapeMarkdownEntities)
    .use(remarkStripTrailingBreaks)
    .use(remarkToMarkdownPlugin)
    .use(remarkAllowAllText)
    .use(createRemarkShortcodeStringifier({ plugins: getEditorComponents() }))
    .use(remarkPlugins);

  /**
   * Transform the MDAST with plugins.
   */
  const processedMdast = processor.runSync(mdast);

  /**
   * Serialize the MDAST to markdown.
   */
  const markdown = processor.stringify(processedMdast).replace(/\r?/g, '');

  /**
   * Return markdown with trailing whitespace removed.
   */
  return trimEnd(markdown);
}

/**
 * Convert Markdown to HTML.
 */
export function markdownToHtml(markdown, { getAsset, resolveWidget, remarkPlugins = [] } = {}) {
  const mdast = markdownToRemark(markdown, remarkPlugins);

  const hast = unified()
    .use(remarkToRehypeShortcodes, { plugins: getEditorComponents(), getAsset, resolveWidget })
    .use(remarkToRehype, { allowDangerousHTML: true })
    .runSync(mdast);

  const html = unified()
    .use(rehypeToHtml, {
      allowDangerousHtml: true,
      allowDangerousCharacters: true,
      closeSelfClosing: true,
      entities: { useNamedReferences: true },
    })
    .stringify(hast);

  return html;
}

/**
 * Deserialize an HTML string to Slate's Raw AST. Currently used for HTML
 * pastes.
 */
export function htmlToSlate(html) {
  const hast = unified().use(htmlToRehype, { fragment: true }).parse(html);

  const mdast = unified()
    .use(rehypePaperEmoji)
    .use(rehypeToRemark, { minify: false })
    .runSync(hast);

  const slateRaw = unified()
    .use(remarkAssertParents)
    .use(remarkPaddedLinks)
    .use(remarkWrapHtml)
    .use(remarkToSlate)
    .runSync(mdast);

  return slateRaw;
}

/**
 * Convert Markdown to Slate's Raw AST.
 */
export function markdownToSlate(markdown, { voidCodeBlock, remarkPlugins = [] } = {}) {
  const mdast = markdownToRemark(markdown, remarkPlugins);

  const slateRaw = unified()
    .use(remarkWrapHtml)
    .use(remarkToSlate, { voidCodeBlock })
    .runSync(mdast);

  return slateRaw.children;
}

/**
 * Convert a Slate Raw AST to Markdown.
 *
 * Requires shortcode plugins to parse shortcode nodes back to text.
 *
 * Note that Unified is not utilized for the conversion from Slate's Raw AST to
 * MDAST. The conversion is manual because Unified can only operate on Unist
 * trees.
 */
export function slateToMarkdown(raw, { voidCodeBlock, remarkPlugins = [] } = {}) {
  const mdast = slateToRemark(raw, { voidCodeBlock });
  const markdown = remarkToMarkdown(mdast, remarkPlugins);
  return markdown;
}
