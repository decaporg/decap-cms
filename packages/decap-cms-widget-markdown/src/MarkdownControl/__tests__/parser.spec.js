import { markdownToSlate } from '../../serializers';

const parser = markdownToSlate;

describe('Compile markdown to Slate Raw AST', () => {
  it('should compile simple markdown', () => {
    const value = `
# H1

sweet body
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "H1",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "text": "sweet body",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile a markdown ordered list', () => {
    const value = `
# H1

1. yo
2. bro
3. fro
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "H1",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "yo",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "bro",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "fro",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
    ],
    "data": Object {
      "start": 1,
    },
    "type": "numbered-list",
  },
]
`);
  });

  it('should compile bulleted lists', () => {
    const value = `
# H1

* yo
* bro
* fro
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "H1",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "yo",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "bro",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "fro",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
    ],
    "data": Object {
      "start": null,
    },
    "type": "bulleted-list",
  },
]
`);
  });

  it('should compile multiple header levels', () => {
    const value = `
# H1

## H2

### H3
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "H1",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "text": "H2",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "H3",
      },
    ],
    "type": "heading-three",
  },
]
`);
  });

  it('should compile horizontal rules', () => {
    const value = `
# H1

---

blue moon
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "H1",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "type": "thematic-break",
  },
  Object {
    "children": Array [
      Object {
        "text": "blue moon",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile horizontal rules', () => {
    const value = `
# H1

---

blue moon
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "H1",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "type": "thematic-break",
  },
  Object {
    "children": Array [
      Object {
        "text": "blue moon",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile soft breaks (double space)', () => {
    const value = `
blue moon  
footballs
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "blue moon",
      },
      Object {
        "children": Array [
          Object {
            "text": "",
          },
        ],
        "data": undefined,
        "type": "break",
      },
      Object {
        "text": "footballs",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile images', () => {
    const value = `
![super](duper.jpg)
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "",
          },
        ],
        "data": Object {
          "alt": "super",
          "title": null,
          "url": "duper.jpg",
        },
        "type": "image",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile code blocks', () => {
    const value = `
\`\`\`javascript
var a = 1;
\`\`\`
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "var a = 1;",
      },
    ],
    "data": Object {
      "lang": "javascript",
      "shortcode": "code-block",
      "shortcodeData": Object {
        "code": "var a = 1;",
        "lang": "javascript",
      },
    },
    "type": "shortcode",
  },
]
`);
  });

  it('should compile nested inline markup', () => {
    const value = `
# Word

This is **some *hot* content**

perhaps **scalding** even
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "Word",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "text": "This is ",
      },
      Object {
        "bold": true,
        "marks": Array [
          Object {
            "type": "bold",
          },
        ],
        "text": "some ",
      },
      Object {
        "bold": true,
        "italic": true,
        "marks": Array [
          Object {
            "type": "bold",
          },
          Object {
            "type": "italic",
          },
        ],
        "text": "hot",
      },
      Object {
        "bold": true,
        "marks": Array [
          Object {
            "type": "bold",
          },
        ],
        "text": " content",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "perhaps ",
      },
      Object {
        "bold": true,
        "marks": Array [
          Object {
            "type": "bold",
          },
        ],
        "text": "scalding",
      },
      Object {
        "text": " even",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile inline code', () => {
    const value = `
# Word

This is some sweet \`inline code\` yo!
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "Word",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "text": "This is some sweet ",
      },
      Object {
        "code": true,
        "marks": Array [
          Object {
            "type": "code",
          },
        ],
        "text": "inline code",
      },
      Object {
        "text": " yo!",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile links', () => {
    const value = `
# Word

How far is it to [Google](https://google.com) land?
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "text": "Word",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "text": "How far is it to ",
      },
      Object {
        "children": Array [
          Object {
            "text": "Google",
          },
        ],
        "data": Object {
          "title": null,
          "url": "https://google.com",
        },
        "type": "link",
      },
      Object {
        "text": " land?",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile plugins', () => {
    const value = `
![test](test.png)

{{< test >}}
`;
    expect(parser(value)).toMatchInlineSnapshot(`
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "",
          },
        ],
        "data": Object {
          "alt": "test",
          "title": null,
          "url": "test.png",
        },
        "type": "image",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "{{< test >}}",
      },
    ],
    "type": "paragraph",
  },
]
`);
  });

  it('should compile kitchen sink example', () => {
    const value = `
# An exhibit of Markdown

This note demonstrates some of what Markdown is capable of doing.

*Note: Feel free to play with this page. Unlike regular notes, this doesn't
automatically save itself.*

## Basic formatting

Paragraphs can be written like so. A paragraph is the basic block of Markdown.
A paragraph is what text will turn into when there is no reason it should
become anything else.

Paragraphs must be separated by a blank line. Basic formatting of *italics* and
**bold** is supported. This *can be **nested** like* so.

## Lists

### Ordered list

1. Item 1 2. A second item 3. Number 3 4. Ⅳ

*Note: the fourth item uses the Unicode character for Roman numeral four.*

### Unordered list

* An item Another item Yet another item And there's more...

## Paragraph modifiers

### Code block

    Code blocks are very useful for developers and other people who look at
    code or other things that are written in plain text. As you can see, it
    uses a fixed-width font.

You can also make \`inline code\` to add code into other things.

### Quote

> Here is a quote. What this is should be self explanatory. Quotes are
automatically indented when they are used.

## Headings

There are six levels of headings. They correspond with the six levels of HTML
headings. You've probably noticed them already in the page. Each level down
uses one more hash character.

### Headings *can* also contain **formatting**

### They can even contain \`inline code\`

Of course, demonstrating what headings look like messes up the structure of the
page.

I don't recommend using more than three or four levels of headings here,
because, when you're smallest heading isn't too small, and you're largest
heading isn't too big, and you want each size up to look noticeably larger and
more important, there there are only so many sizes that you can use.

## URLs

URLs can be made in a handful of ways:

* A named link to MarkItDown. The easiest way to do these is to select what you
* want to make a link and hit \`Ctrl+L\`.  Another named link to
* [MarkItDown](http://www.markitdown.net/) Sometimes you just want a URL like
* <http://www.markitdown.net/>.

## Horizontal rule

A horizontal rule is a line that goes across the middle of the page.

---

It's sometimes handy for breaking things up.

## Images

Markdown can also contain images. I'll need to add something here sometime.

## Finally

There's actually a lot more to Markdown than this. See the official
introduction and syntax for more information. However, be aware that this is
not using the official implementation, and this might work subtly differently
  in some of the little things.
`;
    expect(parser(value)).toMatchSnapshot();
  });
});
