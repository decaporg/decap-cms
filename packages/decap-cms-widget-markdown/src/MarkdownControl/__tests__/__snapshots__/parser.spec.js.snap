// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compile markdown to Slate Raw AST should compile kitchen sink example 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "An exhibit of Markdown",
      },
    ],
    "type": "heading-one",
  },
  Object {
    "children": Array [
      Object {
        "text": "This note demonstrates some of what Markdown is capable of doing.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "italic": true,
        "marks": Array [
          Object {
            "type": "italic",
          },
        ],
        "text": "Note: Feel free to play with this page. Unlike regular notes, this doesn't
automatically save itself.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "Basic formatting",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "Paragraphs can be written like so. A paragraph is the basic block of Markdown.
A paragraph is what text will turn into when there is no reason it should
become anything else.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "Paragraphs must be separated by a blank line. Basic formatting of ",
      },
      Object {
        "italic": true,
        "marks": Array [
          Object {
            "type": "italic",
          },
        ],
        "text": "italics",
      },
      Object {
        "text": " and
",
      },
      Object {
        "bold": true,
        "marks": Array [
          Object {
            "type": "bold",
          },
        ],
        "text": "bold",
      },
      Object {
        "text": " is supported. This ",
      },
      Object {
        "italic": true,
        "marks": Array [
          Object {
            "type": "italic",
          },
        ],
        "text": "can be ",
      },
      Object {
        "bold": true,
        "italic": true,
        "marks": Array [
          Object {
            "type": "italic",
          },
          Object {
            "type": "bold",
          },
        ],
        "text": "nested",
      },
      Object {
        "italic": true,
        "marks": Array [
          Object {
            "type": "italic",
          },
        ],
        "text": " like",
      },
      Object {
        "text": " so.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "Lists",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "Ordered list",
      },
    ],
    "type": "heading-three",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "Item 1 2. A second item 3. Number 3 4. â…£",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
    ],
    "data": Object {
      "start": 1,
    },
    "type": "numbered-list",
  },
  Object {
    "children": Array [
      Object {
        "italic": true,
        "marks": Array [
          Object {
            "type": "italic",
          },
        ],
        "text": "Note: the fourth item uses the Unicode character for Roman numeral four.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "Unordered list",
      },
    ],
    "type": "heading-three",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "An item Another item Yet another item And there's more...",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
    ],
    "data": Object {
      "start": null,
    },
    "type": "bulleted-list",
  },
  Object {
    "children": Array [
      Object {
        "text": "Paragraph modifiers",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "Code block",
      },
    ],
    "type": "heading-three",
  },
  Object {
    "children": Array [
      Object {
        "text": "Code blocks are very useful for developers and other people who look at
code or other things that are written in plain text. As you can see, it
uses a fixed-width font.",
      },
    ],
    "data": Object {
      "lang": null,
      "shortcode": "code-block",
      "shortcodeData": Object {
        "code": "Code blocks are very useful for developers and other people who look at
code or other things that are written in plain text. As you can see, it
uses a fixed-width font.",
        "lang": null,
      },
    },
    "type": "shortcode",
  },
  Object {
    "children": Array [
      Object {
        "text": "You can also make ",
      },
      Object {
        "code": true,
        "marks": Array [
          Object {
            "type": "code",
          },
        ],
        "text": "inline code",
      },
      Object {
        "text": " to add code into other things.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "Quote",
      },
    ],
    "type": "heading-three",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "Here is a quote. What this is should be self explanatory. Quotes are
automatically indented when they are used.",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "quote",
  },
  Object {
    "children": Array [
      Object {
        "text": "Headings",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "There are six levels of headings. They correspond with the six levels of HTML
headings. You've probably noticed them already in the page. Each level down
uses one more hash character.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "Headings ",
      },
      Object {
        "italic": true,
        "marks": Array [
          Object {
            "type": "italic",
          },
        ],
        "text": "can",
      },
      Object {
        "text": " also contain ",
      },
      Object {
        "bold": true,
        "marks": Array [
          Object {
            "type": "bold",
          },
        ],
        "text": "formatting",
      },
    ],
    "type": "heading-three",
  },
  Object {
    "children": Array [
      Object {
        "text": "They can even contain ",
      },
      Object {
        "code": true,
        "marks": Array [
          Object {
            "type": "code",
          },
        ],
        "text": "inline code",
      },
    ],
    "type": "heading-three",
  },
  Object {
    "children": Array [
      Object {
        "text": "Of course, demonstrating what headings look like messes up the structure of the
page.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "I don't recommend using more than three or four levels of headings here,
because, when you're smallest heading isn't too small, and you're largest
heading isn't too big, and you want each size up to look noticeably larger and
more important, there there are only so many sizes that you can use.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "URLs",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "URLs can be made in a handful of ways:",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "A named link to MarkItDown. The easiest way to do these is to select what you",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "text": "want to make a link and hit ",
              },
              Object {
                "code": true,
                "marks": Array [
                  Object {
                    "type": "code",
                  },
                ],
                "text": "Ctrl+L",
              },
              Object {
                "text": ".  Another named link to",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "MarkItDown",
                  },
                ],
                "data": Object {
                  "title": null,
                  "url": "http://www.markitdown.net/",
                },
                "type": "link",
              },
              Object {
                "text": " Sometimes you just want a URL like",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "http://www.markitdown.net/",
                  },
                ],
                "data": Object {
                  "title": null,
                  "url": "http://www.markitdown.net/",
                },
                "type": "link",
              },
              Object {
                "text": ".",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list-item",
      },
    ],
    "data": Object {
      "start": null,
    },
    "type": "bulleted-list",
  },
  Object {
    "children": Array [
      Object {
        "text": "Horizontal rule",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "A horizontal rule is a line that goes across the middle of the page.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "",
      },
    ],
    "type": "thematic-break",
  },
  Object {
    "children": Array [
      Object {
        "text": "It's sometimes handy for breaking things up.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "Images",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "Markdown can also contain images. I'll need to add something here sometime.",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "children": Array [
      Object {
        "text": "Finally",
      },
    ],
    "type": "heading-two",
  },
  Object {
    "children": Array [
      Object {
        "text": "There's actually a lot more to Markdown than this. See the official
introduction and syntax for more information. However, be aware that this is
not using the official implementation, and this might work subtly differently
  in some of the little things.",
      },
    ],
    "type": "paragraph",
  },
]
`;
